Source/DafnyCore/DafnyOptions.cs

```diff
@@ -1103,10 +1156,14 @@ public enum IncludesModes {
       // Boogie also used to set the following options, but does not anymore.
       SetZ3Option("auto_config", "false");
       SetZ3Option("type_check", "true");
-      SetZ3Option("smt.case_split", "3"); // TODO: try removing
       SetZ3Option("smt.qi.eager_threshold", "100"); // TODO: try lowering
       SetZ3Option("smt.delay_units", "true");
-      SetZ3Option("smt.arith.solver", "2");
+
+      if (z3Version is not null && z3Version.Major == 4 && z3Version.Minor == 8 && z3Version.Build == 5) {
+        // These options tend to help with Z3 4.8.5 but hurt with newer versions of Z3.
+        SetZ3Option("smt.case_split", "3");
+        SetZ3Option("smt.arith.solver", "2");
```

Comment:
How much are they hurting? Can you please describe the problem in a separate issue?

 Also consider describing a range of versions in the conditional, rather than an exact version. Perhaps refactor this:
https://github.com/dafny-lang/dafny/pull/3233/files#diff-b3d55f76ae78d57880964c42ed16a37e9e6a8a3a7ca4022cb96367bccda8d667R80

Reply:
My experience is that any version newer than 4.8.5 works better with those options removed. I considered adding code to produce an error if it detects a version _older_ than 4.8.5. If I do that, then the conditional above really should just be for a single version. What do you think?

Reply:
If you add this error if we detect a version older than 4.8.5, then yes that's fine. Where would you add this error?
If you do not add it in this PR, then I would prefer to see a version "<=" instead of just "=".

Reply:
I went for changing it to `<=` for now, since it's a little awkward to a signal an error from that location and I haven't actually seen evidence that versions < 4.8.5 _don't_ work. Before we release 4.0 we may want to test those early versions and have a more complete strategy for how to handle the entire range of available versions.